### PR TITLE
Decouple static workflow graph validation from execution state

### PR DESCRIPTION
**Pull request by krehermann**
The change pre-sages refactoring the workflow validation into `chainlink-common` for use by CLO.

The point of the change is to decouple the notion of valid yaml (that forms a graph) from the notion of an executable workflow whose steps have mutable step that are updated by the engine when it's executed.

To achieve that, we introduce a `StaticWorkflow` that is only the intermediate representation from parsing YAML into a valid graph.

Given this graph, we convert it to the pre-existing notion of a workflow.

The next change will be to move the `StaticWorkflow` into `chainlink-common`, which is the reason that it's public to start.

### CODE REVIEW

**Review by bolekk — DISMISSED** [review]

**Review by ettec — DISMISSED** [review]

File: core/services/workflows/engine.go [hunk 1]
```diff
@@ -260,8 +260,8 @@ func (e *Engine) resumeInProgressExecutions(ctx context.Context) error {
 // and `scheduledExecution` for targets. If we don't have the necessary
 // config to initialize a scheduledExecution for a target, we'll fallback to
 // using `immediateExecution`.
-func (e *Engine) initializeExecutionStrategy(step *step) error {
-	if step.executionStrategy != nil {
+func (e *Engine) initializeExecutionStrategy(s *step) error {
```
Comment (HenryNguyen5) [inline core/services/workflows/engine.go:263]: I find this change quite confusing, it's easy for me to get lost when there's single character vars in a large function. Can we keep it as-is?

File: core/services/workflows/engine.go [hunk 2]
```diff
@@ -540,10 +540,10 @@ func (e *Engine) handleStepUpdate(ctx context.Context, stepUpdate store.Workflow
 	return nil
 }
 
-func (e *Engine) queueIfReady(state store.WorkflowExecution, step *step) {
+func (e *Engine) queueIfReady(state store.WorkflowExecution, s *step) {
```
Comment (HenryNguyen5) [inline core/services/workflows/engine.go]: I find this change quite confusing, it's easy for me to get lost when there's single character vars in a large function. Can we keep it as-is?

File: core/services/workflows/models.go [hunk 3]
```diff
@@ -106,17 +106,42 @@ func (w *workflow) dependents(start string) ([]*step, error) {
 	return steps, nil
 }
 
-// step wraps a stepDefinition with additional context for dependencies and execution
+// step wraps a Vertex with additional context for dependencies and execution
```
Comment (HenryNguyen5) [inline core/services/workflows/models.go]: Nice!

Comment (HenryNguyen5) [reply core/services/workflows/models.go]: Vertex actually contains the `dependencies` field though, does this comment need to be modified?

File: core/services/workflows/models.go [hunk 4]
```diff
@@ -106,17 +106,42 @@ func (w *workflow) dependents(start string) ([]*step, error) {
...
+type StaticWorkflow struct {
+	ID string
+	graph.Graph[string, *Vertex]
+
+	Triggers []*StepDefinition
+
+	Spec *WorkflowSpec
+}
+
+func (w *StaticWorkflow) Hasher() func(s *Vertex) string {
```
Comment (HenryNguyen5) [inline core/services/workflows/models.go]: What's the purpose behind this change?

File: core/services/workflows/models.go [hunk 5]
```diff
@@ -106,17 +106,42 @@ func (w *workflow) dependents(start string) ([]*step, error) {
...
+	return func(s *Vertex) string {
+		return s.Ref
```
Comment (HenryNguyen5) [inline core/services/workflows/models.go]: This isn't a hash, could we have a different function name?

File: core/services/workflows/models.go [hunk 6]
```diff
@@ -106,17 +106,42 @@ func (w *workflow) dependents(start string) ([]*step, error) {
...
+	}
+}
+
+func (v *Vertex) Hash() string {
+	return v.Ref
```
Comment (HenryNguyen5) [inline core/services/workflows/models.go:137]: This isn't a hash, could we have a different function name?

File: core/services/workflows/models.go [hunk 7]
```diff
@@ -200,16 +233,68 @@ func Parse(yamlWorkflow string) (*workflow, error) {
...
 		Graph:    g,
-		triggers: triggerSteps,
+		Triggers: triggerSteps,
 	}
 	return wf, err
 }
+
+// createExecutableWorkflow converts a StaticWorkflow to an executable workflow
+// by adding metadata to the vertices that is owned by the workflow runtime.
+func createExecutableWorkflow(wf2 *StaticWorkflow) (*workflow, error) {
```
Comment (HenryNguyen5) [inline core/services/workflows/models.go]: I would still consider this static evaluation, it's still a pure function. I think it'd be good to have this within `chainlink-common` too.

Comment (HenryNguyen5) [reply core/services/workflows/models.go]: The big difference between this function and `ParseStaticWorkflow` is that one  ensures the workflow matches our schema, the other ensures that the workflow is a valid graph (in a workflow sense)

**Review by HenryNguyen5 — CHANGES_REQUESTED** [review]

File: core/services/workflows/engine.go (hunk 1 above)
Comment (krehermann) [reply core/services/workflows/engine.go:263]: the linter was unhappy about shadowing. 

i could change it to `stp` or any other name than `step`. what do you suggest?

File: core/services/workflows/models.go (hunk 3 above)
Comment (krehermann) [reply core/services/workflows/models.go]: ah, yes. need to change the comment. good catch. thanks

File: core/services/workflows/models.go (hunk 4 above)
Comment (krehermann) [reply core/services/workflows/models.go]: the graph lib takes a func that compute the id of a vertex. that lib calls it a hash, and the existing code call it stepHash, and i just reused the name.

the computation need to be common btw the intermediate rep and the workflow so that it's possible to copy the graph structure.

File: core/services/workflows/models.go (hunk 7 above)
Comment (krehermann) [reply core/services/workflows/models.go]: the workflow is mutated by the engine. i'm opening to naming suggestions. i don't want to change `workflow` though because one of the goals is minimal surgery to decouple the mutable state from the static state.

Comment (krehermann) [reply core/services/workflows/models.go]: re: moving this to common. i don't want to do that do because the mutable state is really tightly coupled to the engine. We'd have to prompt all the private fields to public, and it don't really make sense to separate it from the engine impl. (i start down that road before discovering that its possible to decouple the static and mutable state via an intermediate representation.)

File: core/services/workflows/engine.go (hunk 1 above)
Comment (krehermann) [reply core/services/workflows/engine.go:263]: changed and seems like the linter in my IDE is too sensitive.

**Review by HenryNguyen5 — APPROVED** [review]

**Review by bolekk — APPROVED** [review]
